feat: 🚀 1단계 - 상품 관리 기능

hongmoSung opened the pull request:
## 코드리뷰 요청 드립니다.

### 요구사항  

- [x] 상품 목록 페이지 연동
- [x] 상품 관리 CRUD API 작성
- [x] 관리자 도구 페이지 연동

hongmoSung committed:
feat: 🚀 1단계 - 상품 관리 기능

- [x] 상품 목록 페이지 연동
- [x] 상품 관리 CRUD API 작성
- [x] 관리자 도구 페이지 연동

Gomding commented on src/main/java/cart/presentation/ViewController.java:
```diff
@@ -0,0 +1,33 @@
+
+import java.util.List;
+
+@Controller
+public class ViewController {
```
View Controller 와 Rest API Controller 를 분리해주셨네요 👍

Gomding commented on src/main/java/cart/data/ProductRepositoryImpl.java:
```diff
@@ -0,0 +1,74 @@
+import java.util.Map;
+
+@Repository
+public class ProductRepositoryImpl implements ProductRepository {
```
구현체에 이름 짓기 애매할 때는 Impl 이라고 짓는 방법도 있지만
현재는 `JDBC` 라는것을 사용하는 Repository 이니 이것이 구현체 이름에 드러나면 어떨까요? 🤔

Gomding commented on src/main/java/cart/data/ProductRepositoryImpl.java:
```diff
@@ -0,0 +1,74 @@
+@Repository
+public class ProductRepositoryImpl implements ProductRepository {
+
+    private final SimpleJdbcInsert simpleJdbcInsert;
```
SimpleJdbcInsert 를 사용해주셨군요 👍

Gomding commented on src/main/java/cart/data/ProductRepositoryImpl.java:
```diff
@@ -0,0 +1,74 @@
+            int productPrice = rs.getInt("product_price");
+            String productImageUrl = rs.getString("product_image_url");
+            return new CartProduct(productId, productName, productPrice, productImageUrl);
+        });
```
RowMapper 코드가 아래 `getProductById` 메서드에서도 동일하게 나타나고 있네요!
아래와 같이 선언해서 중복을 제거하는건 어떻게 생각하시나요?

```java
@Repository
public class ProductRepositoryImpl implements ProductRepository {
    private final SimpleJdbcInsert simpleJdbcInsert;
    private final RowMapper<CartProduct> rowMapper = (rs, rowNum) -> {
            long productId = rs.getLong("product_id");
            String productName = rs.getString("product_name");
            int productPrice = rs.getInt("product_price");
            String productImageUrl = rs.getString("product_image_url");
            return new CartProduct(productId, productName, productPrice, productImageUrl);
        }
}
```

Gomding commented on src/main/java/cart/business/AdminService.java:
```diff
@@ -0,0 +1,42 @@
+package cart.business;
```
코멘트 남길 곳이 없어 여기에 남깁니다 ㅠㅠ

test 중 깨지는 테스트가 있어 확인 부탁드립니다!
크게 두 가지 이유 때문에 테스트가 깨지고 있는것으로 보여요

1. 테스트 **메서드마다** `data.sql` 이라는 스크립트를 실행
2. `/products` path를 가지는 GET method 요청을 API 가 허용하지 않음 (405 status code)

<img width="308" alt="image" src="https://github.com/next-step/spring-cart/assets/57378410/ed167186-b230-41c7-9bf1-7e420b7b35ba">

Gomding commented on src/main/java/cart/business/AdminService.java:
```diff
@@ -0,0 +1,42 @@
+        productRepository.addProduct(cartProduct);
+    }
+
+    public List<ViewProduct> getViewProducts() {
```
메서드 이름이 `getViewProducts` 일 필요가 있을지 의문이 들었어요!
특정 DTO 에 한정된 느낌이 듭니다 😢 

Service 는 Product 전부 조회해서 응답해준다는것만 알면 될것같아요 😄

Gomding commented on src/main/java/cart/presentation/AdminController.java:
```diff
@@ -0,0 +1,31 @@
+
+@RequestMapping("/products")
+@RestController
+public class AdminController {
```
Product 에 대한 CUD 기능을 담당하는데 이름이 `AdminController` 라서 직관적이지 않은것같아요 🤔 

홍모님은 어떻게 생각하시나요?

Gomding commented on src/main/java/cart/business/AdminService.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.stream.Collectors;
+
+@Service
+public class AdminService {
```
이 서비스도 내용물은 Product 에 대한 기능인데 `AdminService` 라서 이름이 직관적이지 않은것 같아요!

Gomding commented on src/main/java/cart/data/entity/CartProduct.java:
```diff
@@ -0,0 +1,55 @@
+
+import java.time.LocalDateTime;
+
+public class CartProduct {
```
entity 패키지의 CartProduct 와 domain 패키지의 Product 는 어떤 차이가 있는지 궁금해요!

hongmoSung replied:
entity는 db용으로 하고 domain은 service 레이어 용도로 사용하려고 나눴는데...

사실 이부분에 대해서 리뷰어님의 조언을 듣고 싶습니다!

Gomding replied:
분리하고자 하는 의도가 궁금하군요 :)

db 용 엔티티 객체와 도메인 객체를 나눴을 때 어떤 장점이 있어 분리해주셨을까요?

hongmoSung replied:
음... 사실... Spring-racing 프로젝트 과제를 하면서 레이어별로 분리하는 리펙토링 과정에서 나누면서
이번 과제에도 자연스럽게 아무런 고려없이 작성하였습니다 😅
다만 앤티티 객체와 도메인 객체를 나누었을때 경험은 DB에 의존하지 않고 테스트코드를 작성하는데 편했던 경험이 있었습니다..

Gomding commented on src/main/java/cart/business/AdminService.java:
```diff
@@ -0,0 +1,42 @@
+
+    public void deleteProduct(long id) {
+        productRepository.removeProduct(id);
+    }
```
현재 삭제해야할 id를 가진 상품이 저장소에 존재하지 않으면 어떻게 동작하고 있을까요?

Gomding commented on src/main/java/cart/data/ProductRepositoryImpl.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    @Override
+    public CartProduct getProductById(long id) {
+        JdbcTemplate jdbcTemplate = simpleJdbcInsert.getJdbcTemplate();
+        String sql = "select * from CART_PRODUCT where product_id = ?";
+        return jdbcTemplate.queryForObject(sql, (rs, rowNum) -> {
```
`getProductById` 호출 시 조회되는 객체가 없다면 어떻게 되고 있을까요!

조회결과 1건도 존재하지 않는다면 예외가 발생하고 있을 것 같습니다 🤔 
Optional 타입을 반환해서 값이 있는지 확인하는 것을 호출하는 쪽에 맡겨보면 어떨까요?

[Java-Optional-바르게-쓰기](http://homoefficio.github.io/2019/10/03/Java-Optional-%EB%B0%94%EB%A5%B4%EA%B2%8C-%EC%93%B0%EA%B8%B0/)

Gomding reviewed (changes requested):
안녕하세요 홍모님!
장바구니 미션 빠르게 진행해주셨네요 :)
몇 가지 코멘트 남겼으니 확인 부탁드립니다! 🙏 

궁금한 점 있으시면 언제든 편하게 DM 주세요 😄

hongmoSung committed:
refactor: 1단계 상품관리 코드리뷰

- amdin -> product 네이밍 변경
- select Optional 처리
- delete 실패 케이스 처리
- test 코드 실패 처리

Gomding commented on src/main/java/cart/data/ProductJdbcRepository.java:
```diff
@@ -1,21 +1,25 @@
 
 @Repository
-public class ProductRepositoryImpl implements ProductRepository {
+public class ProductJdbcRepository implements ProductRepository {
```
Jdbc 를 사용하는 Repository 인것이 드러나도록 변경해주셨네요 :)

Gomding commented on src/main/java/cart/data/ProductJdbcRepository.java:
```diff
@@ -0,0 +1,85 @@
+        JdbcTemplate jdbcTemplate = simpleJdbcInsert.getJdbcTemplate();
+        String sql = "select * from CART_PRODUCT where product_id = ?";
+        CartProduct cartProduct = null;
+        try {
+            cartProduct = jdbcTemplate.queryForObject(sql, this.cartProductMapper(), id);
+        } catch (Exception e) {
+            logger.info("sql 오류 -> {}", e.getMessage());
+        }
+        return Optional.ofNullable(cartProduct);
```
아래와 같이 반환해보는 것은 어떠신가요!
예외가 발생했을 때 비어있는 Optional 값을 전달하고
호출하는 쪽에서는 Optional 이 비어있는지 확인하고 사용하도록요 😄 

```suggestion
        try {
            return Optional.of(jdbcTemplate.queryForObject(sql, this.cartProductMapper(), id));
        } catch (Exception e) {
            return Optional.emtpy();
        }
```

Gomding replied:
아래와 같이 반환해보는 것은 어떠신가요!
예외가 발생했을 때 비어있는 Optional 값을 전달하고
호출하는 쪽에서는 Optional 이 비어있는지 확인하고 사용하도록요 😄 

```suggestion
        try {
            return Optional.of(jdbcTemplate.queryForObject(sql, this.cartProductMapper(), id));
        } catch (Exception e) {
            return Optional.emtpy();
        }
```

Gomding commented on src/main/java/cart/data/ProductJdbcRepository.java:
```diff
@@ -0,0 +1,85 @@
+        CartProduct cartProduct = null;
+        try {
+            cartProduct = jdbcTemplate.queryForObject(sql, this.cartProductMapper(), id);
+        } catch (Exception e) {
```
Exception 을 catch 하면 정말 SQL 이 에러가 난 상황 까지도 흡수해버린 다는 단점이 있을것같아요
(예외가 발생해야하는데 Optional 타입으로 반환해버림)

조회했을 때 1건도 존재하지 않을 때 또는 1건 이상 존재할 때 발생하는 예외가 있을거에요 :)
JdbcTemplate 의 queryForObject 라는 메서드 내부를 확인해볼까요?

```java
// JdbcTemplate
	@Override
	@Nullable
	public <T> T queryForObject(String sql, RowMapper<T> rowMapper, @Nullable Object... args) throws DataAccessException {
		List<T> results = query(sql, args, new RowMapperResultSetExtractor<>(rowMapper, 1));
		return DataAccessUtils.nullableSingleResult(results);
	}

// DataAccessUtils.java
	@Nullable
	public static <T> T nullableSingleResult(@Nullable Collection<T> results) throws IncorrectResultSizeDataAccessException {
		// This is identical to the requiredSingleResult implementation but differs in the
		// semantics of the incoming Collection (which we currently can't formally express)
		if (CollectionUtils.isEmpty(results)) {
			throw new EmptyResultDataAccessException(1);
		}
		if (results.size() > 1) {
			throw new IncorrectResultSizeDataAccessException(1, results.size());
		}
		return results.iterator().next();
	}
```

Gomding commented on src/main/java/cart/data/ProductJdbcRepository.java:
```diff
@@ -69,6 +67,19 @@ public void updateProduct(CartProduct cartProduct) {
         JdbcTemplate jdbcTemplate = simpleJdbcInsert.getJdbcTemplate();
         String sql = "delete from CART_PRODUCT where product_id = ?";
-        jdbcTemplate.update(sql, id);
+        int updateCount = jdbcTemplate.update(sql, id);
+        if (updateCount != 1) {
+            throw new RuntimeException(id + "에 해당하는 상품을 삭제하는데 오류가 발생하였습니다.");
+        }
```
delete 시 DAO 는 updateCount 만을 반환하고
예외처리는 호출하는 쪽에서 해보면 어떨까요?

DAO 는 가능하면 바깥에서 어떻게 사용될 지에 관심을 가지지 않도록 구현하는것이 좋습니다 :)
어떤 로직에서는 삭제할 때 대상이 없어도 예외가 발생하지 않아야할 수도 있겠죠?

Gomding commented on src/main/java/cart/presentation/ViewController.java:
```diff
@@ -11,22 +11,22 @@
 public class ViewController {
 
-    private final AdminService adminService;
+    private final ProductService adminService;
```
필드 이름이 안바뀐듯 하군요! 😄

Gomding commented on src/main/resources/data.sql:
```diff
@@ -0,0 +1,24 @@
+create table CART_PRODUCT
```
해당 테이블은 Product 라는 정보에 집중된 테이블이니 `PRODUCT`로 이름을 지어도 괜찮아보여요!

Gomding commented on src/main/resources/data.sql:
```diff
@@ -0,0 +1,24 @@
+create table CART_PRODUCT
+(
+    product_id        int auto_increment primary key,
+    product_name      varchar(50)  not null,
+    product_price     int          not null,
+    product_image_url varchar(500) not null,
+    created_at        datetime     not null
```
이 부분은 매우 취향적인 부분이지만

product 라는 부분은 테이블 이름에 표현이 됐으므로
컬럼 이름에는 `product_` 라는 prefix 는 없어도 테이블 정보를 이해하는데 문제가 없어보여요 :)

+) image_url 은 varchar 대신 text 타입으로 지정해줘도 괜찮겠어요! (URL 이 얼마나 길어질지 가늠이 안되는군요 😢 )

Gomding commented on src/main/java/cart/data/ProductJdbcRepository.java:
```diff
@@ -0,0 +1,85 @@
+import org.springframework.jdbc.core.simple.SimpleJdbcInsert;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public class ProductJdbcRepository implements ProductRepository {
```
Spring 프레임워크를 사용한 테스트도 작성해보면 어떨까요?

간단하게 `@JdbcTest` 라는 애너테이션을 사용해서 Repository 테스트부터 작성해보시면 좋겠어요!
https://www.baeldung.com/spring-jdbctemplate-testing

+) 더 나아가서 Service , Controller 테스트도 작성해보시면 좋겠군요 😄 
Controller 테스트는 `ProductIntegrationTest` 에 샘플 테스트가 있으니 다른 path 들도 테스트를 작성해볼 수 있겠네요!

Gomding reviewed (approved):
안녕하세요 홍모님!
피드백 반영 빠르게 해주셨네요 👍 
이번 단계는 여기서 merge 하겠습니다!
추가로 남긴 코멘트는 다음 단계에서 함께 고민해주세요 🙏 
(1단계 PR에 대댓글로 질문이나 답변 주시면 확인하겠습니다 😄 )

궁금한 점 있으시면 언제든 편하게 DM 남겨주세요!
2단계도 화이팅입니다 🔥
